datafusion/functions/src/math/log.rs

```diff
@@ -0,0 +1,259 @@
+            DataType::Float64 => match base {
+                ColumnarValue::Scalar(ScalarValue::Float32(Some(base))) => {
+                    Arc::new(make_function_scalar_inputs!(x, "x", Float64Array, {
+                        |value: f64| f64::log(value, base as f64)
+                    }))
+                }
+                ColumnarValue::Array(base) => Arc::new(make_function_inputs2!(
+                    x,
+                    base,
+                    "x",
+                    "base",
+                    Float64Array,
+                    { f64::log }
+                )),
+                _ => {
+                    return exec_err!("log function requires a scalar or array for base")
+                }
+            },
+
+            DataType::Float32 => match base {
+                ColumnarValue::Scalar(ScalarValue::Float32(Some(base))) => {
+                    Arc::new(make_function_scalar_inputs!(x, "x", Float32Array, {
+                        |value: f32| f32::log(value, base)
+                    }))
+                }
+                ColumnarValue::Array(base) => Arc::new(make_function_inputs2!(
+                    x,
+                    base,
+                    "x",
+                    "base",
+                    Float32Array,
+                    { f32::log }
+                )),
+                _ => {
+                    return exec_err!("log function requires a scalar or array for base")
+                }
+            },
+            other => {
+                return exec_err!("Unsupported data type {other:?} for function log")
+            }
+        };
+
+        Ok(ColumnarValue::Array(arr))
+    }
+
+    /// Simplify the `log` function by the relevant rules:
+    /// 1. Log(a, 1) ===> 0
+    /// 2. Log(a, Power(a, b)) ===> b
+    /// 3. Log(a, a) ===> 1
+    fn simplify(
+        &self,
+        args: Vec<Expr>,
+        info: &dyn SimplifyInfo,
+    ) -> Result<ExprSimplifyResult> {
+        let mut number = &args[0];
+        let mut base =
+            &Expr::Literal(ScalarValue::new_ten(&info.get_data_type(number)?)?);
+        if args.len() == 2 {
+            base = &args[0];
+            number = &args[1];
+        }
+
+        match number {
+            Expr::Literal(value)
+                if value == &ScalarValue::new_one(&info.get_data_type(number)?)? =>
+            {
+                Ok(ExprSimplifyResult::Simplified(Expr::Literal(
+                    ScalarValue::new_zero(&info.get_data_type(base)?)?,
+                )))
+            }
+            Expr::ScalarFunction(ScalarFunction {
+                func_def: ScalarFunctionDefinition::UDF(fun),
+                args,
+            }) if base == &args[0]
+                && fun
+                    .as_ref()
+                    .inner()
+                    .as_any()
+                    .downcast_ref::<PowerFunc>()
+                    .is_some() =>
+            {
+                Ok(ExprSimplifyResult::Simplified(args[1].clone()))
```

Comment:
It would be really nice to avoid this clone (I am working to reduce the number of times we need to clone `Expr`s during planning).

Reply:
Makes sense. I'll open up a follow-up PR with the suggested changes after this is merged.